ui/src/map/layers/utils/publication-candidate-highlight-utils.ts

```diff
@@ -0,0 +1,474 @@
+    },
+    changeType: explicitness,
+});
+
+const equalsRange = (min: number, max: number, range: Range<number>): boolean =>
+    min === range.min && max === range.max;
+
+/**
+ * Divides points into groups by given m-value ranges and spaces between ranges.
+ * Each group created by m-value range is marked as an explicit change, groups between
+ * m-value ranges are marked as implicit changes.
+ *
+ * Each returned group (or index range) contains at least two points and ranges are inclusive,
+ * e.g. 1..3, 3..10, 10..12 etc.
+ */
+const splitByRanges = (points: AlignmentPoint[], mRanges: Range<number>[]): PointRange[] => {
+    const explicitChangeIndexRanges: Range<number>[] = mRanges
+        .sort((range1, range2) => range1.min - range2.min)
+        .map((mRange) => findSubRange(points, mRange))
+        .filter(filterNotEmpty);
+
+    const mergedExplicitIndexRanges = mergeOverlappingRanges(explicitChangeIndexRanges);
+
+    const rangeEndpoints = [
+        0,
+        ...mergedExplicitIndexRanges.map((range) => [range.min, range.max]).flat(),
+        lastIndex(points),
+    ].filter(filterUnique);
+    return init(rangeEndpoints).map((currentMin, index) => {
+        const currentMax = expectDefined(rangeEndpoints[index + 1]);
+
+        return mergedExplicitIndexRanges.some((range) => equalsRange(currentMin, currentMax, range))
+            ? createPointRange(currentMin, currentMax, ChangeExplicitness.EXPLICIT)
+            : createPointRange(currentMin, currentMax, ChangeExplicitness.IMPLICIT);
+    });
+};
+
+const getHighlightColor = (
+    stage: PublicationStage,
+    explicitness: ChangeExplicitness,
+    operation: Operation,
+): string => {
+    if (stage === PublicationStage.STAGED) {
+        return explicitness === ChangeExplicitness.EXPLICIT
+            ? mapStyles.previewHighlightStagedExplicit
+            : mapStyles.previewHighlightStagedImplicit;
+    } else if (operation === 'DELETE') {
+        return explicitness === ChangeExplicitness.EXPLICIT
+            ? mapStyles.previewHighlightDeletedExplicit
+            : mapStyles.previewHighlightDeletedImplicit;
+    } else {
+        return explicitness === ChangeExplicitness.EXPLICIT
+            ? mapStyles.previewHighlightModifiedExplicit
+            : mapStyles.previewHighlightModifiedImplicit;
+    }
+};
+
+const getHighlightZIndex = (
+    operation: Operation,
+    assetType: DraftChangeType,
+    stage: PublicationStage,
+    explicitness: ChangeExplicitness,
+): number => {
+    let zIndex = HIGHEST_HIGHLIGHT_Z_INDEX;
+
+    // Unstaged changes go below staged changes
+    if (stage === PublicationStage.UNSTAGED) zIndex -= 8;
+    // Implicit changes go below explicit changes
+    if (explicitness === ChangeExplicitness.IMPLICIT) zIndex -= 4;
+    // Points go below LineStrings
+    if (!isLineStringType(assetType)) zIndex -= 2;
+    // Deletions go below other operations
+    if (operation === 'DELETE') zIndex -= 1;
+
+    return zIndex;
+};
```

Comment:
Tämä refaktoroitu toimimaan toisin päin (otetaan korkein Z-indeksi, joka on nyt vakioitu, ja miinustellaan siitä ominaisuuksien mukaan.) Täten z-indekseille saadaan mahdollisesti helpommin ymmärrettävissä oleva arvoalue, kun myös alignment-viivojen z-indeksit saadaan sidottua tähän (ks. r84-86)